utils/filename.go

```diff
@@ -0,0 +1,60 @@
+package utils
+
+import (
+	"fmt"
+	"regexp"
+	"strings"
+)
+
+const lowerHex = "0123456789abcdef"
+
+var allChars = regexp.MustCompile(`.`)
+
+// Symbols that are never allowed in file names on at least one platform
+// Note that we include '%', which is allowed, but we need to use it as an escape character
+var illegalSymbols = regexp.MustCompile(`[?<>*|\\/%:"^[:cntrl:]]`)
+
+// These are reserved words and symbols which cannot be a file name, with or without a file extension
+// Mainly special Windows devices
+// See https://learn.microsoft.com/en-us/windows/win32/fileio/naming-a-file#naming-conventions
+var reservedWords = regexp.MustCompile(`^aux|^com[0-9¹²³]|^con|^lpt[0-9¹²³]|^nul|^prn`)
+
+// Symbols that cannot (or should not) end a file name
+// Windows does not allow file names to end in a '.' or ' '
+var illegalSuffix = regexp.MustCompile(`[. ]$`)
+
+// Encode percent encodes a filename string, to minimize cross-platform compatibility issues related to case handling,
+// allowed characters, and reserved words
+func Encode(str string) (string, error) {
+	if len(str) == 0 {
+		return str, nil
+	}
+	var filename = str
+	filename = illegalSymbols.ReplaceAllStringFunc(filename, escapeAll)
+	filename = allChars.ReplaceAllStringFunc(filename, escapeUpper)
+	filename = illegalSuffix.ReplaceAllStringFunc(filename, escapeAll)
+
+	parts := strings.SplitN(filename, ".", 2)
+	parts[0] = reservedWords.ReplaceAllStringFunc(parts[0], escapeAll)
+
+	return strings.Join(parts, "."), nil
+}
+
+// escapeAll converts every byte in string s to a percent encoded string
+// similar to net/url methods, but simplified to remove any decision logic, and using lowercase letters
+func escapeAll(s string) string {
+	var buf strings.Builder
+
+	for i := 0; i < len(s); i++ {
+		buf.WriteString(fmt.Sprintf("%%%02x", s[i]))
```

Comment:
Ah yeah you're right we need to iterate on bytes not runes.

That said, you can use Fprintf which is going to write more directly to the string builder

```suggestion
		fmt.Fprintf(&buf, "%%%02x", s[i])
```

Reply:
Fprintf returns an error, which is hard to deal with here. Unless you want me to just ignore it? The error seems to be about failing to write to the buffer, which shouldn't be an issue when the buffer is a strings.Builder.

Reply:
the error is caused by the underlying error returned by the io.Writer; in our case that's the string builder so it can be safely ignored

Reply:
(btw buf.WriteString also returns an error; it does so in order to implement  the https://pkg.go.dev/io#StringWriter interface)

Reply:
Huh, I wonder why Goland complained about the unhandled error from one, but not the other. 🤷🏻‍♀️

Anyway, I updated to use Fprintf.